File: packages/llm-obs-node-packages/shared-infra/src/http/resilience/standard-circuit-breaker.ts
/**
 * @file standard-circuit-breaker.ts
 * @description Standard Bounded LRU Circuit Breaker State Machine.
 * 
 * ALGORITHM & STATE MACHINE SPECIFICATION:
 * 1. State Machine Transitions:
 *    - CLOSED (Normal): Allows network calls. Increments failure counter on error.
 *      Transitions to OPEN if failures >= failureThreshold (default: 5).
 *    - OPEN (Tripped): Fast-fails requests with 503 without socket initiation.
 *      Transitions to HALF_OPEN after cooldown period (cooldownMs, default: 3600000ms / 1hr).
 *    - HALF_OPEN (Probe): Allows 1 trial request.
 *      If trial succeeds -> CLOSED (reset).
 *      If trial fails -> OPEN (reset cooldown).
 * 2. Memory Protection (Bounded LRU Capacity):
 *    - Bounded to maxEntries (default: 1000) using Map insertion ordering LRU eviction.
 */

import { HTTP_CONSTANTS } from "../constants";
import { deriveRouteTemplate } from "../utils/http-utils";

export interface CircuitState {
  state: "CLOSED" | "OPEN" | "HALF_OPEN";
  failures: number;
  lastFailureTime?: number;
  nextAttemptTime?: number;
}

export class StandardCircuitBreaker {
  private readonly states = new Map<string, CircuitState>();

  constructor(
    private readonly failureThreshold = 5,
    private readonly cooldownMs = 3600000,
    private readonly maxEntries = 1000
  ) {}

  public getCircuitKey(tenantId: string, routeTemplateOrUrl: string): string {
    const route = deriveRouteTemplate(routeTemplateOrUrl);
    return `${tenantId}:${route}`;
  }


  public canExecute(circuitKey: string): boolean {
    const state = this.states.get(circuitKey);
    if (!state || state.state === HTTP_CONSTANTS.CIRCUIT_CLOSED) {
      return true;
    }

    const now = Date.now();
    if (state.state === HTTP_CONSTANTS.CIRCUIT_OPEN) {
      if (state.nextAttemptTime && now >= state.nextAttemptTime) {
        state.state = HTTP_CONSTANTS.CIRCUIT_HALF_OPEN;
        return true;
      }
      return false;
    }

    if (state.state === HTTP_CONSTANTS.CIRCUIT_HALF_OPEN) {
      return true;
    }

    return true;
  }

  public onSuccess(circuitKey: string): void {
    const state = this.states.get(circuitKey);
    if (state) {
      state.state = HTTP_CONSTANTS.CIRCUIT_CLOSED as "CLOSED";
      state.failures = 0;
      state.nextAttemptTime = undefined;
    }
  }

  public onFailure(circuitKey: string): void {
    const now = Date.now();
    let state = this.states.get(circuitKey);

    if (!state) {
      this.evictIfFull();
      state = { state: HTTP_CONSTANTS.CIRCUIT_CLOSED as "CLOSED", failures: 0 };
      this.states.set(circuitKey, state);
    }

    state.failures++;
    state.lastFailureTime = now;

    if (state.failures >= this.failureThreshold || state.state === HTTP_CONSTANTS.CIRCUIT_HALF_OPEN) {
      state.state = HTTP_CONSTANTS.CIRCUIT_OPEN as "OPEN";
      state.nextAttemptTime = now + this.cooldownMs;
    }
  }

  public getState(circuitKey: string): CircuitState | undefined {
    return this.states.get(circuitKey);
  }

  private evictIfFull(): void {
    if (this.states.size >= this.maxEntries) {
      const oldestKey = this.states.keys().next().value;
      if (oldestKey) {
        this.states.delete(oldestKey);
      }
    }
  }
}

File: packages/llm-obs-python-packages/python-shared/src/python_shared/http/resilience/standard_circuit_breaker.py
import time
import threading
from typing import Dict, Optional
from python_shared.http.constants import HTTP_CONSTANTS

class CircuitState:
    def __init__(self, state: str = HTTP_CONSTANTS.CIRCUIT_CLOSED, failures: int = 0):
        self.state = state
        self.failures = failures
        self.last_failure_time: Optional[float] = None
        self.next_attempt_time: Optional[float] = None

class StandardCircuitBreaker:
    def __init__(
        self,
        failure_threshold: int = 5,
        cooldown_ms: float = 3600000.0,
        max_entries: int = 1000
    ):
        self.failure_threshold = failure_threshold
        self.cooldown_ms = cooldown_ms
        self.max_entries = max_entries
        self._states: Dict[str, CircuitState] = {}
        self._lock = threading.RLock()

    def get_circuit_key(self, tenant_id: str, route_template_or_url: str) -> str:
        return f"{tenant_id}:{route_template_or_url}"

    def can_execute(self, circuit_key: str) -> bool:
        with self._lock:
            state = self._states.get(circuit_key)
            if not state or state.state == HTTP_CONSTANTS.CIRCUIT_CLOSED:
                return True

            now = time.time() * 1000.0
            if state.state == HTTP_CONSTANTS.CIRCUIT_OPEN:
                if state.next_attempt_time and now >= state.next_attempt_time:
                    state.state = HTTP_CONSTANTS.CIRCUIT_HALF_OPEN
                    return True
                return False

            if state.state == HTTP_CONSTANTS.CIRCUIT_HALF_OPEN:
                return True

            return True

    def on_success(self, circuit_key: str) -> None:
        with self._lock:
            state = self._states.get(circuit_key)
            if state:
                state.state = HTTP_CONSTANTS.CIRCUIT_CLOSED
                state.failures = 0
                state.next_attempt_time = None

    def on_failure(self, circuit_key: str) -> None:
        now = time.time() * 1000.0
        with self._lock:
            state = self._states.get(circuit_key)
            if not state:
                self._evict_if_full()
                state = CircuitState(state=HTTP_CONSTANTS.CIRCUIT_CLOSED, failures=0)
                self._states[circuit_key] = state

            state.failures += 1
            state.last_failure_time = now

            if state.failures >= self.failure_threshold or state.state == HTTP_CONSTANTS.CIRCUIT_HALF_OPEN:
                state.state = HTTP_CONSTANTS.CIRCUIT_OPEN
                state.next_attempt_time = now + self.cooldown_ms

    def get_state(self, circuit_key: str) -> Optional[CircuitState]:
        with self._lock:
            return self._states.get(circuit_key)

    def _evict_if_full(self) -> None:
        if len(self._states) >= self.max_entries:
            oldest_key = next(iter(self._states.keys()), None)
            if oldest_key:
                del self._states[oldest_key]
